pandas/_libs/algos_rank_helper.pxi.in

```diff
@@ -8,110 +8,104 @@ WARNING: DO NOT edit .pxi FILE directly, .pxi is generated from .pxi.in
+ctypedef fused rank_t:
+    object
+    float64_t
+    uint64_t
+    int64_t
 
 
 @cython.wraparound(False)
 @cython.boundscheck(False)
-def rank_1d_{{dtype}}({{ctype}}[:] in_arr, ties_method='average',
-                      ascending=True, na_option='keep', pct=False):
+def rank_1d(rank_t[:] in_arr, ties_method='average',
+            ascending=True, na_option='keep', pct=False):
     """
     Fast NaN-friendly version of scipy.stats.rankdata
     """
 
     cdef:
         Py_ssize_t i, j, n, dups = 0, total_tie_count = 0, non_na_idx = 0
 
-        {{if dtype == 'object'}}
-        ndarray sorted_data, values
-        {{else}}
-        ndarray[{{ctype}}] sorted_data, values
-        {{endif}}
+        ndarray[rank_t] sorted_data, values
 
         ndarray[float64_t] ranks
         ndarray[int64_t] argsorted
         ndarray[uint8_t, cast=True] sorted_mask
 
-        {{if dtype == 'uint64'}}
-        {{ctype}} val
-        {{else}}
-        {{ctype}} val, nan_value
-        {{endif}}
+        rank_t val, nan_value
 
         float64_t sum_ranks = 0
         int tiebreak = 0
         bint keep_na = 0
-        bint isnan
+        bint isnan, condition
         float64_t count = 0.0
+
     tiebreak = tiebreakers[ties_method]
 
-    {{if dtype == 'float64'}}
-    values = np.asarray(in_arr).copy()
-    {{elif dtype == 'object'}}
-    values = np.array(in_arr, copy=True)
+    if rank_t is float64_t:
+        values = np.asarray(in_arr).copy()
+    elif rank_t is object:
+        values = np.array(in_arr, copy=True)
 
-    if values.dtype != np.object_:
-        values = values.astype('O')
-    {{else}}
-    values = np.asarray(in_arr)
-    {{endif}}
+        if values.dtype != np.object_:
+            values = values.astype('O')
+    else:
+        values = np.asarray(in_arr)
 
     keep_na = na_option == 'keep'
 
-    {{if dtype == 'object'}}
-    mask = missing.isnaobj(values)
-    {{elif dtype == 'float64'}}
-    mask = np.isnan(values)
-    {{elif dtype == 'int64'}}
-    mask = values == NPY_NAT
+    if rank_t is object:
+        mask = missing.isnaobj(values)
+    elif rank_t is float64_t:
+        mask = np.isnan(values)
+    elif rank_t is int64_t:
+        mask = values == NPY_NAT
 
-    # create copy in case of NPY_NAT
-    # values are mutated inplace
-    if mask.any():
-        values = values.copy()
-    {{endif}}
+        # create copy in case of NPY_NAT
+        # values are mutated inplace
+        if mask.any():
+            values = values.copy()
 
     # double sort first by mask and then by values to ensure nan values are
     # either at the beginning or the end. mask/(~mask) controls padding at
     # tail or the head
-    {{if dtype != 'uint64'}}
-    if ascending ^ (na_option == 'top'):
-        nan_value = {{pos_nan_value}}
-        order = (values, mask)
+    if rank_t is not uint64_t:
+        if ascending ^ (na_option == 'top'):
+            if rank_t is object:
```

Comment:
Not necessarily in scope for this but would be nice if there was a better way to do this via fused types (a dict or inline func?)

Reply:
Yah I've been giving that some thought.  I think we can do that with something like:

```
cdef inline rank_t get_thing(rank_t val):  # <-- totally ignore `val`, just care about rank_t
    if rank_t is object: ...

def actual_func(rank_t ...):
    cdef:
          rank_t val, thing
    
    thing = get_thing(val)
```